solidity/contracts/contracts/QuantumGravityBridge.sol

```diff
@@ -0,0 +1,314 @@
+            revert InsufficientVotingPower();
+        }
+    }
+
+    /// @notice This updates the validator set by checking that the validators
+    /// in the current validator set have signed off on the new validator set.
+    /// The signatures supplied are the signatures of the current validator set
+    /// over the checkpoint hash generated from the new validator set. Anyone
+    /// can call this function, but they must supply valid signatures of the
+    /// current validator set over the new validator set.
+    /// @param _newValidatorSetHash The hash of the new validator set
+    /// @param _newNonce The new nonce.
+    /// @param _currentValidatorSet The current validator set.
+    /// @param _sigs Signatures.
+    function updateValidatorSet(
+        uint256 _newNonce,
+        uint256 _newPowerThreshold,
+        bytes32 _newValidatorSetHash,
+        Validator[] calldata _currentValidatorSet,
+        Signature[] calldata _sigs
+    ) external {
+        // CHECKS
+
+        uint256 currentNonce = s_lastValidatorSetNonce;
+        uint256 currentPowerThreshold = s_powerThreshold;
+
+        // Check that the valset nonce is greater than the old one.
+        if (_newNonce <= currentNonce) {
+            revert InvalidValidatorSetNonce();
+        }
+
+        // Check that current validators and signatures are well-formed.
+        if (_currentValidatorSet.length != _sigs.length) {
+            revert MalformedCurrentValidatorSet();
+        }
+
+        // Check that the supplied current validator set matches the saved checkpoint
+        bytes32 currentValsetHash = keccak256(abi.encode(_currentValidatorSet));
+        if (
+            domainSeparateValidatorSetHash(BRIDGE_ID, currentNonce, currentPowerThreshold, currentValsetHash) !=
+            s_lastValidatorSetCheckpoint
+        ) {
+            revert SuppliedValidatorSetInvalid();
+        }
+
+        // Check that enough current validators have signed off on the new validator set
+        bytes32 newCheckpoint = domainSeparateValidatorSetHash(
+            BRIDGE_ID,
+            currentNonce,
+            _newPowerThreshold,
+            _newValidatorSetHash
+        );
+        checkValidatorSignatures(_currentValidatorSet, _sigs, newCheckpoint, currentPowerThreshold);
+
+        // EFFECTS
+
+        s_lastValidatorSetCheckpoint = newCheckpoint;
+        s_powerThreshold = _newPowerThreshold;
+        s_lastValidatorSetNonce = _newNonce;
+
+        // LOGS
+
+        emit ValidatorSetUpdatedEvent(_newNonce, _newPowerThreshold, _newValidatorSetHash);
+    }
+
+    /// @notice Relays a batch of Celestia -> Ethereum messages. Anyone can
+    // call this function, but they must supply valid signatures of the current
+    // validator set over the batch.
+    /// @param _nonce The message root nonce.
+    /// @param _messageRoot The Merkle root of messages.
+    /// @param _currentValidatorSet The current validator set.
+    /// @param _sigs Signatures.
```

Comment:
Signatures of what? Above it says batch. But it's not clear what the batch really is here.

From reading the code they actually sign the result of:
`keccak256(abi.encode(_bridge_id, methodName, _nonce, _messageRoot))`

Reply:
Will improve docs here.

Reply:
This still needs to be addressed.

Reply:
Improved in cdb3dce.